pmd-java/src/main/java/net/sourceforge/pmd/lang/java/oom/ClassStats.java

```diff
@@ -0,0 +1,177 @@
+     */
+    void addOperation(String name, OperationSignature sig) {
+        if (!operations.containsKey(sig)) {
+            operations.put(sig, new HashSet<OperationStats>());
+        }
+        operations.get(sig).add(new OperationStats(name));
+    }
+
+    /**
+     * Adds a field to the class.
+     *
+     * @param name The qualified name of the field
+     * @param sig  The signature of the field
+     */
+    void addField(String name, FieldSignature sig) {
+        if (!fields.containsKey(sig)) {
+            fields.put(sig, new HashSet<String>());
+        }
+        fields.get(sig).add(name);
+    }
+
+    /**
+     * Checks whether the class declares an operation by the name given which is covered by the
+     * signature mask.
+     *
+     * @param name The name of the operation to look for
+     * @param mask The mask covering accepted signatures
+     *
+     * @return True if the class declares an operation by the name given which is covered by the signature mask, false
+     * otherwise.
+     */
+    boolean hasMatchingSig(String name, OperationSigMask mask) {
+        // Indexing on signatures optimises this type of request
+        for (OperationSignature sig : operations.keySet()) {
+            if (mask.covers(sig)) {
+                if (operations.get(sig).contains(new OperationStats(name))) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Checks whether the class declares a field by the name given which is covered by the
+     * signature mask.
+     *
+     * @param name The name of the operation to look for
+     * @param mask The mask covering accepted signatures
+     *
+     * @return True if the class declares a field by the name given which is covered by the signature mask, false
+     * otherwise.
+     */
+    boolean hasMatchingSig(String name, FieldSigMask mask) {
+        for (FieldSignature sig : fields.keySet()) {
+            if (mask.covers(sig)) {
+                if (fields.get(sig).contains(name)) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Computes the value of a metric for an operation.
+     *
+     * @param key   The operation metric for which to find a memoized result.
+     * @param node  The AST node of the operation.
+     * @param name  The name of the operation.
+     * @param force Force the recomputation. If unset, we'll first check for a memoized result.
+     *
+     * @return The result of the computation, or {@code Double.NaN} if it couldn't be performed.
+     */
+    double compute(OperationMetricKey key, ASTMethodOrConstructorDeclaration node, String name, boolean force) {
+        // TODO:cf maybe find a way to optimise this
+        for (OperationSignature sig : operations.keySet()) {
+            for (OperationStats stats : operations.get(sig)) {
+                if (stats.equals(name)) {
```

Comment:
This equals looks strange: OperationStats != String... A normal, well-behaving equals method should always return false for this comparison, since the classes don't match. We might need an additional method like `equalsName(...)`.